### PR TITLE
8329555: Crash in intrinsifying heap-based MemorySegment Vector store/loads

### PR DESCRIPTION
**Pull request by jatin-bhateja**
- Problem is related to bi-morphic inlining of ms.unsafeBase() in presence of multiple receiver type profiles (OfFloat, OfByte) which results into formation of an abstract type phi node at JVM state convergence points.
- Due to this, for mismatched memory segment access, inline expander is unable to infer the element type of array wrapped within the memory segment and this results into an assertion failure while computing the source lane count.
- For non-masked mismatched memory segment vector read/write accesses, intrinsification can continue with unknown backing storage type and compiler can skip inserting explicit reinterpretation IR after loading from or before storing to backing storage which is mandatory for semantic correctness of big-endian memory layout.
- For vector write access, this may prevent value forwarding, which may result into subsequent redundant vector loads from same address, but preventing intensification failure will offset that cost.

Kindly review and share your feedback.

Best Regards,
Jatin

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8329555](https://bugs.openjdk.org/browse/JDK-8329555): Crash in intrinsifying heap-based MemorySegment Vector store/loads (**Bug** - P3)

### Reviewers
 * [Sandhya Viswanathan](https://openjdk.org/census#sviswanathan) (@sviswa7 - **Reviewer**)
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18749/head:pull/18749` \
`$ git checkout pull/18749`

Update a local copy of the PR: \
`$ git checkout pull/18749` \
`$ git pull https://git.openjdk.org/jdk.git pull/18749/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18749`

View PR using the GUI difftool: \
`$ git pr show -t 18749`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18749.diff">https://git.openjdk.org/jdk/pull/18749.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18749#issuecomment-2050927348)

### CODE REVIEW

**Comment by jatin-bhateja** [comment]
Graph shape describing the problem.
image

File: src/hotspot/share/opto/vectorIntrinsics.cpp [hunk 1]
```diff
@@ -1037,7 +1037,13 @@ bool LibraryCallKit::inline_vector_mem_operation(bool is_store) {
   const bool needs_cpu_membar = is_mixed_access || is_mismatched_access;
 
   bool mismatched_ms = from_ms->get_con() && !is_mask && arr_type != nullptr && arr_type->elem()->array_element_basic_type() != elem_bt;
-  BasicType mem_elem_bt = mismatched_ms ? arr_type->elem()->array_element_basic_type() : elem_bt;
+  BasicType mem_elem_bt = LITTLE_ENDIAN_ONLY(elem_bt) BIG_ENDIAN_ONLY(arr_type->elem()->array_element_basic_type());
```
Comment (sviswa7) [inline src/hotspot/share/opto/vectorIntrinsics.cpp:1040]: mismatched_ms check is missing now for BIG_ENDIAN_ONLY path

**Review by sviswa7 — APPROVED** [review]
Looks good to me.

**Comment by sviswa7** [comment]
It will be good to get another review as well.
/Reviewers 2

**Comment by jatin-bhateja** [comment]
> Looks good.
> 
> It would be interesting to look on splitting intrinsic through Phi so that we can generate vector on each branch. In separate RFE.

Thanks @vnkozlov , vector Intrinsic expect to receive absolute lane and element type in order to associate correct ideal types with generated vector IR.   Splitting intrinsic by pushing it backwards across phi may lead to multiple inline expansions of an intrinsic given that multiple arguments (element type, lane count, mask etc..) may be phi nodes, drawback could be code bloating which may have other side effects, since these Phi node are generated based on profiles. I agree with you, it will interesting to study its effects in a separate RFE.

**Pull request closed without merging**